File: solution/241. Different Ways to Add Parentheses/241.cpp
class Solution {
public:
    vector<int> diffWaysToCompute(string expression) {
        vector<int> res = dfs(expression, 0, expression.size() - 1);
        return res;
    }
    
private:
    vector<int> dfs(const string& E, int start, int end) {
        int i = start, num = 0;
        vector<int> res;
        for (int i = start; i <= end; ++i) {
            if (isdigit(E[i])) continue;
            
            auto L = dfs(E, start, i - 1);
            auto R = dfs(E, i + 1, end);
            for (auto& l : L)
                for (auto& r : R)
                    res.push_back(calc(l, r, E[i]));
        }
        if (res.empty())
            res.push_back(stoi(E.substr(start, end - start + 1)));
        return res;
    }
    
    int calc(int a, int b, char op) {
        if (op == '+') return a + b;
        if (op == '-') return a - b;
        if (op == '*') return a * b;
        return 0;
    }
};

File: solution/241. Different Ways to Add Parentheses/241.py
class Solution:
    def diffWaysToCompute(self, expression: str) -> List[int]:
        
        def dfs(start, end):
            res = []
            for i in range(start, end + 1):
                if expression[i].isdigit():
                    continue
                
                L = dfs(start, i - 1)
                R = dfs(i + 1, end)
                for l in L:
                    for r in R:
                        res.append(self.calc(l, r, expression[i]))
            if not res:
                res.append(int(expression[start:end+1]))
            return res
        
        return dfs(0, len(expression) - 1)
        
    
    def calc(self, a, b, op):
        if op == '+':
            return a + b
        if op == '-':
            return a - b
        if op == '*':
            return a * b
        return 0
